### PR TITLE
[MachineSink] Clear kill flags of sunk addressing mode registers

### PR DESCRIPTION
**Pull request by momchil-velikov**
When doing sink-and-fold, the MachineSink clears the "killed" flags of the operands of the sunk (and deleted) instruction. However, this is not always sufficient. In some cases we can create the new load/store instruction with operands other than the ones present in the deleted instruction. One such example is folding a zero word extend into a memory load on AArch64. The zero-extend is represented by a pair of instructions - `MOV` (i.e. `ORRwrs`) followed by a `SUBREG_TO_REG`. The `SUBREG_TO_REG` is deleted (it is the sunk instruction), but the new load instruction mentions operands "killed" in the `MOV`, which is no longer correct.

To fix this, clear the "killed" flags of the registers paritipating in the adressing mode.

### CODE REVIEW

File: llvm/lib/CodeGen/MachineSink.cpp [hunk 1]
```diff
@@ -530,6 +525,14 @@ bool MachineSinking::PerformSinkAndFold(MachineInstr &MI,
     } else {
```
Comment (davemgreen) [inline llvm/lib/CodeGen/MachineSink.cpp:525]: Do we need to worry about the isCopy path above too?

File: llvm/test/CodeGen/AArch64/sink-and-fold-clear-kill-flags.mir [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+  maxCallFrameSize: 0
+  cvBytesOfCalleeSavedRegisters: 0
+  hasOpaqueSPAdjustment: false
+  hasVAStart:      false
+  hasMustTailInVarArgFunc: false
+  hasTailCall:     false
+  localFrameSize:  0
+  savePoint:       ''
+  restorePoint:    ''
+fixedStack:      []
```
Comment (davemgreen) [inline llvm/test/CodeGen/AArch64/sink-and-fold-clear-kill-flags.mir:79]: You might find you can remove a lot of this

File: llvm/lib/CodeGen/MachineSink.cpp (hunk 1 above)
Comment (momchil-velikov) [reply llvm/lib/CodeGen/MachineSink.cpp:525]: No, the copies are ok. We have a chain/tree of copies:
```
r1 = copy r0
...
rn = copy rn-1
x = op a, b, rn
```
and then we create `x = op a, b, r0` and all the copies, including the first one are deleted in this function, so it doesn't matter what flags are left in the short time between creation of the new instruction and the deletion of the old one.

The case for zext on AArch64 is a little bit different in that we don't delete the `ORRWrs` , we delete only the `SUBREG_TO_REG`, but at the same time we use an operand from `ORRWrs`.

In principle, we could try to delete dead instructions going up to the beginning of the block, but the correctness check is
a bit more complex and we run the DeadMachineInstrucitonElim pass shortly after anyway.

File: llvm/test/CodeGen/AArch64/sink-and-fold-clear-kill-flags.mir (hunk 2 above)
Comment (momchil-velikov) [reply llvm/test/CodeGen/AArch64/sink-and-fold-clear-kill-flags.mir]: Done.

File: llvm/lib/CodeGen/MachineSink.cpp (hunk 1 above)
Comment (davemgreen) [reply llvm/lib/CodeGen/MachineSink.cpp:525]: I was thinking of this, if we sink op1 pasted the killed a
```
x = op1 a, b
y = op2 killed a
z = copy x
```

Comment (momchil-velikov) [reply llvm/lib/CodeGen/MachineSink.cpp:525]: Yes, you are right.

Comment (momchil-velikov) [reply llvm/lib/CodeGen/MachineSink.cpp:533]: Fixed.

**Review by davemgreen — APPROVED** [review]
LGTM

**Pull request merged**